libcontainer/nsenter/cloned_binary.c

```diff
@@ -415,29 +423,89 @@ static int try_bindfd(void)
 	 */
-	ret = -EPERM;
-	if (mount("/proc/self/exe", template, "", MS_BIND, "") < 0)
+	if ((ret = socketpair(AF_LOCAL, SOCK_STREAM, 0, pipe)) < 0)
 		goto out;
-	if (mount("", template, "", MS_REMOUNT | MS_BIND | MS_RDONLY, "") < 0)
-		goto out_umount;
 
-	/* Get read-only handle that we're sure can't be made read-write. */
-	ret = open(template, O_PATH | O_CLOEXEC);
+	is_child = setjmp(env);
+	if (is_child) {
+		close(pipe[0]);
+		fd = -1;
+		/*
+		 * For obvious reasons this won't work in rootless mode because we haven't
+		 * created a userns -- but getting that to work will be a bit complicated and
+		 * it's only worth doing if someone actually needs it.
+		 */
+		if ((ret = mount("none", "/", NULL, MS_PRIVATE | MS_REC, NULL)) < 0)
+			goto out_child;
+		/*
+		 * The kernel refuses to bind-mount from the magic symlink when the process has
+		 * been cloned (or unshared) into a new mount namespace, at least on Linux 4.4.
+		 */
+		char linkbuf[PATH_MAX + 1] = { 0 };
+		ssize_t linkpathlen = readlink("/proc/self/exe", linkbuf, sizeof(linkbuf));
+		if (linkpathlen == sizeof(linkbuf)) {
+			/*
+			 * The link path is longer than PATH_MAX, and the contents of
+			 * linkbuf might have been truncated. A truncated path could
+			 * happen to be a valid path to a different file, which could
+			 * allow for local privilege escalation if we were to exec it.
+			 * The mount syscall doesn't accept paths longer than PATH_MAX,
+			 * anyway.
+			 */
+			ret = -ENAMETOOLONG;
+			goto out_child;
+		}
+		linkbuf[linkpathlen] = '\0';
+		if ((ret = mount(linkbuf, template, "", MS_BIND, "")) < 0)
+			goto out_child;
+		if ((ret = mount("", template, "", MS_REMOUNT | MS_BIND | MS_RDONLY, "")) < 0)
+			goto out_child;
+
+		/* Get read-only handle that we're sure can't be made read-write. */
+		fd = open(template, O_PATH | O_CLOEXEC);
+		if (fd < 0) {
+			ret = fd;
+			goto out_child;
+		}
 
-out_umount:
-	/*
-	 * Make sure the MNT_DETACH works, otherwise we could get remounted
-	 * read-write and that would be quite bad (the fd would be made read-write
-	 * too, invalidating the protection).
-	 */
-	if (umount2(template, MNT_DETACH) < 0) {
-		if (ret >= 0)
-			close(ret);
-		ret = -ENOTRECOVERABLE;
+		/*
+		 * Make sure the MNT_DETACH works, otherwise we could get remounted
+		 * read-write and that would be quite bad (the fd would be made read-write
+		 * too, invalidating the protection).
+		 */
+		if (umount2(template, MNT_DETACH) < 0) {
+			ret = -ENOTRECOVERABLE;
+			goto out_child;
+		}
+		send_fd(pipe[1], fd);
+out_child:
+		if (fd >= 0)
+			close(fd);
+		close(pipe[1]);
+		exit(-ret);
+	} else {
+		int cpid = clone_longjmp(&env, 1, CLONE_NEWNS | SIGCHLD);
+		if (cpid < 0) {
+			ret = -errno;
+			goto out;
+		}
+		int wstatus = 0;
+		if ((ret = waitpid(cpid, &wstatus, 0)) < 0)
+			goto out;
+		if ((ret = -WEXITSTATUS(wstatus)))
+			goto out;
+		fd = receive_fd(pipe[0]);
+		if ((ret = fcntl(ret, F_SETFD, O_PATH | O_CLOEXEC)) < 0)
+			goto out;
+		ret = fd;
```

Comment:
It looks like a mixup of `ret` and `fd` here. For one thing, passing `ret` to `fcntl` makes no sense; neither does `ret = fd` (or I'm missing something).